lib/chef/provisioning/aws_driver/aws_provider.rb

```diff
@@ -22,8 +23,184 @@ def region
     new_resource.driver.aws_config.region
   end
 
-  def driver
-    new_resource.driver
+  #
+  # Return the damned value from the block, not whatever weirdness converge_by
+  # normally returns.
+  #
+  def converge_by(*args, &block)
+    result = nil
+    super(*args) do
+      result = block.call
+    end
+    result
   end
+
+  action :create do
+    #
+    # If the user specified an ID, get the object for it, and fail if it does not exist.
+    #
+    desired_driver = new_resource.driver
+    desired_id = new_resource.public_send(new_resource.class.aws_id_attribute) if new_resource.class.aws_id_attribute
+    if desired_id
+      aws_object = new_resource.class.get_aws_object(desired_id, resource: new_resource)
+    end
+
+    #
+    # If Chef has already associated the object with an AWS ID, check if it's
+    # the same as the desired AWS ID.
+    #
+    if new_resource.is_a?(AWSResourceWithEntry)
+      entry_driver, entry_id, entry = new_resource.get_id_from_managed_entry
+      if entry_id
+        if desired_id
+
+          #
+          # We have both a desired ID and an entry ID.  Find out whether they
+          # match and warn if they don't (because we're going to reassociate and
+          # update the *desired* AWS thing.).
+          #
+          if desired_driver.driver_url == entry_driver.driver_url && desired_id == entry_id
+            Chef::Log.debug "#{new_resource.to_s} is already associated with #{entry_id} in #{entry_driver.driver_url}"
+          else
+            Chef::Log.warn "#{new_resource.to_s} is currently associated with #{entry_id} in #{entry_driver.driver_url}, but the desired ID is #{desired_id} in #{new_resource.driver.driver_url}!  Will associate with new desired ID #{desired_id}."
```

Comment:
The ID changes in save_managed_entry, a bit below.  `desired_id` and `desired_driver` will be used to update `entry`.